projects/ADuCM3029_demo_cn0414/src/platform_drivers.c

```diff
@@ -0,0 +1,996 @@
+	if(gpio_init_flag == 0) {
+		ret = adi_gpio_Init(mem_gpio_handler, ADI_GPIO_MEMORY_SIZE);
+		if(ret != ADI_GPIO_SUCCESS)
+			goto error;
+	}
+
+	/* Driving strength must be enabled for I2C pins */
+	ret = adi_gpio_DriveStrengthEnable(ADI_GPIO_PORT0,
+			ADI_GPIO_PIN_4 | ADI_GPIO_PIN_5, true);
+	if(ret != ADI_GPIO_SUCCESS)
+		goto error;
+
+	/* If no more GPIOs free driver memory */
+	if (gpio_init_flag == 0) {
+		ret = adi_gpio_UnInit();
+		if(ret != ADI_GPIO_SUCCESS)
+			goto error;
+	}
+
+	/* Open Master Device */
+	ret = adi_i2c_Open(dev->id, master_i2c_device_memory, ADI_I2C_MEMORY_SIZE,
+			&master_i2c_dev);
+	if(ret != ADI_I2C_SUCCESS)
+		goto error;
+
+	/* Set bit rate */
+	ret = adi_i2c_SetBitRate(master_i2c_dev, dev->max_speed_hz);
+	if(ret != ADI_I2C_SUCCESS)
+		goto error;
+
+	*desc = dev;
+
+	return ret;
+error:
+	free(dev);
+
+	return ret;
+}
+
+/**
+ * Free the resources allocated by i2c_init().
+ *
+ * @param [in] desc - The I2C descriptor.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t i2c_remove(i2c_desc *desc)
+{
+	/* Free memory for device descriptor */
+	free(desc);
+
+	/* Close I2C device */
+	return adi_i2c_Close(master_i2c_dev);
+}
+
+/**
+ * Write data to a slave device.
+ *
+ * @param [in] desc         - The I2C descriptor.
+ * @param [in] data         - Buffer that stores the transmission data.
+ * @param [in] bytes_number - Number of bytes to write.
+ * @param [in] stop_bit     - Stop condition control.
+ *                          Example: 0 - A stop condition will not be generated;
+ *                                   1 - A stop condition will be generated.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t i2c_write(i2c_desc *desc, uint8_t *data, uint8_t bytes_number,
+		uint8_t stop_bit)
+{
+	ADI_I2C_TRANSACTION trans;
+	uint32_t error;
+
+	trans.bReadNotWrite = false;
+	if(stop_bit == 0)
+		trans.bRepeatStart = true;
+	else
+		trans.bRepeatStart = false;
+	trans.nDataSize = bytes_number;
+	trans.nPrologueSize = 0;
+	trans.pData = data;
+	trans.pPrologue = NULL;
+
+	return adi_i2c_ReadWrite(master_i2c_dev, &trans, &error);
+}
+
+/**
+ * Read data from a slave device.
+ *
+ * @param [in] desc         - The I2C descriptor.
+ * @param [out] data        - Buffer that will store the received data.
+ * @param [in] bytes_number - Number of bytes to read.
+ * @param [in] stop_bit     - Stop condition control.
+ *                          Example: 0 - A stop condition will not be generated;
+ *                                   1 - A stop condition will be generated.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t i2c_read(i2c_desc *desc, uint8_t *data, uint8_t bytes_number,
+		uint8_t stop_bit)
+{
+	int32_t ret = 0;
+	ADI_I2C_TRANSACTION trans;
+	uint32_t error;
+
+	trans.bReadNotWrite = true;
+	if(stop_bit == 0)
+		trans.bRepeatStart = true;
+	else
+		trans.bRepeatStart = false;
```

Comment:
This is correct, but I would write as bellow, since it saves a couple of lines:
```
trans.bReadNotWrite = true;
trans.bRepeatStart = true;
if (stop_bit)
        trans.bRepeatStart = false;
```
Feel free to disagree :)